src/host_beta/messageRouter.ts

```diff
@@ -0,0 +1,169 @@
+                this.recordPerformanceHistogram(histogramName, duration);
+                return true;
+            case 'reportError':
+                const [type, message, stack, filename, sourceUrl, lineno, colno] = args;
+                this.reportError(type, message, stack, filename, sourceUrl, lineno, colno);
+                return true;
+            case 'sendMessageToBackend':
+                const [cdpMessage] = args;
+                this.sendMessageToBackend(cdpMessage);
+                return true;
+            default:
+                // TODO: handle other types of messages from devtools
+                return false;
+        }
+    }
+
+    onMessageFromChannel(e: WebviewEvent, args: string): boolean {
+        if (e !== 'websocket') {
+            return false;
+        }
+        const { event, message } = JSON.parse(args) as {event: WebSocketEvent, message: string};
+        this.fireWebSocketCallback(event, message);
+        return true;
+    }
+
+    private sendReady() {
+        // Inform the extension we are ready to receive messages
+        encodeMessageForChannel(msg => vscode.postMessage(msg, '*'), 'ready');
+    }
+
+    private recordEnumeratedHistogram(actionName: string, actionCode: number, _bucketSize: number): void {
+        // Inform the extension of the DevTools telemetry event
+        this.sendTelemetry({
+            data: actionCode,
+            event: 'enumerated',
+            name: actionName,
+        });
+    }
+
+    private recordPerformanceHistogram(histogramName: string, duration: number): void {
+        // Inform the extension of the DevTools telemetry event
+        this.sendTelemetry({
+            data: duration,
+            event: 'performance',
+            name: histogramName,
+        });
+    }
+
+    private reportError(
+        type: string,
+        message: string,
+        stack: string,
+        filename: string,
+        sourceUrl: string,
+        lineno: number,
+        colno: number): void {
+        // Package up the error info to send to the extension
+        const data = { message, stack, filename, sourceUrl, lineno, colno };
+
+        // Inform the extension of the DevTools telemetry event
+        this.sendTelemetry({
+            data,
+            event: 'error',
+            name: type,
+        });
+    }
+
+    private openInEditor(url: string, line: number, column: number, ignoreTabChanges: boolean): void {
+        // Forward the data to the extension
+        const request: IOpenEditorData = { column, line, url, ignoreTabChanges };
+        encodeMessageForChannel(msg => vscode.postMessage(msg, '*'), 'openInEditor', request);
+    }
+
+    private openInNewTab(url: string): void {
+        encodeMessageForChannel(msg => vscode.postMessage(msg, '*'), 'openUrl', {url});
+    }
+
+    private sendMessageToBackend(message: string): void {
+        encodeMessageForChannel(msg => vscode.postMessage(msg, '*'), 'websocket', { message });
```

Comment:
if we already have a hold for the webviewWindow (passed in the constructor) should we be able to use that instead of using the acquireVSCodeApi?

Reply:
No, these messages need to be passed to the main extension, not the webview window